registry/tokentransport.go

```diff
@@ -70,11 +67,11 @@ func (t *TokenTransport) auth(authService *authService) (string, *http.Response,
 	if err != nil {
 		return "", nil, err
 	}
+	defer response.Body.Close()
 
 	if response.StatusCode != http.StatusOK {
 		return "", response, err
 	}
```

Comment:
This defer looks to potentially break current usages of RoundTrip when there is an `authReq` error.

Looks like before the change, when the `authReq` fails, the response is bubbled up and returned in `RoundTrip()`, where it becomes the responsibility of the consumer to handle the cleanup.

With the defer the consumer will now get a response with a closed body and be unable to read it.

Does it make more sense to do something similar to this instead:

```suggestion
	if err != nil {
	        _ = response.Body.Close()
		return "", nil, err
	}

	if response.StatusCode != http.StatusOK {
		return "", response, err
	}
	defer response.Body.Close()
```

Reply:
This is essentially what we had before. I think the problem is typically when `err != nil` is returned, it is expected that the client can ignore the other returned values (in this case, the response). I actually don't really like that this library keeps returning the response even when `err != nil`.

I took a look at https://pkg.go.dev/net/http#Client.Do, and it says `On error, any Response can be ignored.`, which tells me the response's body *should* be closed (guess that's not specified here). I also checked [the implementation](https://cs.opensource.google/go/go/+/refs/tags/go1.22.2:src/net/http/client.go;l=259), and I see when `RoundTrip` returns an error the response is ignored

Reply:
I can see a mixture of checking if `resp.Body == nil` before closing and just simply ignoring once we know `err == nil` used throughout this library. A bit odd

Reply:
you know what, this PR may be unnecessary because `response.StatusCode != http.StatusOK` was already covered by the original 

```
if authResp != nil && authResp.Body != nil {
	_ = authResp.Body.Close()
}
```

the fact the original author returned `err` instead of explicitly setting it to `nil` was confusing. If you agree, I'm going to close this PR, as I believe we already have what we want

Reply:
actually instead of closing, i'll keep the PR but make it clear this `err` is actually `nil`